Reworked version of Config Broker

PhilDay-CT opened the pull request:
The is a rework of the Configuration Broker Demo, in a much more realistic form.

- Incoming values are now serialised JSON, as might be expected from an MQTT broker
- Values are more realistic rather than purely abstract test values
- JSON parsing is the main focus of the sandbox containment

Note: The JSON parser is built around a couple of files copied in from FreeRTOS

PhilDay-CT committed:
Reworked version of Config Broker

davidchisnall commented on configuration_broker/config_broker/config_broker.cc:
```diff
@@ -0,0 +1,343 @@
+	std::atomic<uint32_t> version;    // version - used as a futex
+	void                 *data;       // current value
+	const char           *name;       // name
+	size_t                size;       // size of the created object
+	uint32_t              minTicks;   // Min system ticks between updates
```
These possibly want to be `const` if they should be set when a config item is registered and then not updated.

PhilDay-CT replied:
They're set when the parser is registered, but the object can be created before then if a consumer tries to read a value first.

davidchisnall commented on configuration_broker/config_broker/config_broker.cc:
```diff
@@ -0,0 +1,343 @@
+// Debugging can be enable with "xmake --config --debug-config_broker=true"
+using Debug = ConditionalDebug<DEBUG_CONFIG_BROKER, "Config Broker">;
+
+// Internal view of a Config Item.
```
Please can you use doc comments (`///` or `/** ... */`) for the comments that explain what something is (so IDE integration works better)?

davidchisnall commented on configuration_broker/config_broker/config_broker.cc:
```diff
@@ -0,0 +1,343 @@
+// deleting an item). The total number of items is limited by the
+// the set of static sealed capabilities.
+//
+InternalConfigitem *configData = nullptr;
```
If this is used only from this compilation unit, it's a good idea to place it in an anonymous namespace so that the compiler knows that it can see all accesses.  The same applies to functions.

davidchisnall commented on configuration_broker/config_broker/config_broker.cc:
```diff
@@ -0,0 +1,343 @@
+// create one.  Use a LockGuard to protect against two
+// threads trying to create the same item.
+//
+FlagLock lockFindOrCreate;
```
If this is used only in `find_or_create_config`, can it be a `static` in that function?

davidchisnall commented on configuration_broker/config_broker/config_broker.cc:
```diff
@@ -0,0 +1,343 @@
+
+	// Allocate a Config object
+	InternalConfigitem *c =
+	  static_cast<InternalConfigitem *>(malloc(sizeof(InternalConfigitem)));
```
```suggestion
	InternalConfigitem *c = new(std::nothrow) InternalConfigitem();
```

davidchisnall commented on configuration_broker/config_broker/config_broker.cc:
```diff
@@ -0,0 +1,343 @@
+	// use the first thread that tries to load a value to
+	// make a cross-comparement call to parserInit()
+	//
+	static bool ParsersLoaded = false;
+
+	if (!ParsersLoaded)
+	{
+		parserInit();
+		ParsersLoaded = true;
+	}
+
```
This might work better using C++ static initialisation.  Something like:
```c++
static bool parsersLoaded = parserInit();
```

The compiler will then insert code that runs `parserInit` exactly once, in a thread-safe way.

PhilDay-CT replied:
I'm guessing because parserInit is a cross compartment call if I make it static then I get a linker error ?

```
error: ld.lld: error: undefined symbol: __library_export_libcalls___cxa_guard_acquire
>>> referenced by __cpp_config_broker.cc.cc
>>>               build/cheriot/cheriot/release/config_broker.compartment:()

ld.lld: error: undefined symbol: __library_export_libcalls___cxa_guard_release
>>> referenced by __cpp_config_broker.cc.cc
>>>               build/cheriot/cheriot/release/config_broker.compartment:()
```

davidchisnall commented on configuration_broker/config_broker/config_broker.cc:
```diff
@@ -0,0 +1,343 @@
+	size_t                size;       // size of the created object
+	uint32_t              minTicks;   // Min system ticks between updates
+	uint64_t              nextUpdate; // Time of next valid update
+	FlagLock              lock;       // lock to prevent concurrent changes
```
Given how it's used, this probably wants to be a priority-inheriting flag lock.

davidchisnall commented on configuration_broker/config_broker/config_broker.cc:
```diff
@@ -0,0 +1,343 @@
+	c->nextUpdate = tick + c->minTicks;
+
+	// Allocate heap space for the new value
+	void *newData = malloc(c->size);
```
It would be good to use a `std::unique_ptr` here so that you don't have to remember to free.

davidchisnall commented on configuration_broker/config_broker/config_broker.cc:
```diff
@@ -0,0 +1,343 @@
+	  roData.permissions().without(CHERI::Permission::Store);
+
+	// Keep track of the old value so we can free it
+	auto oldData = c->data;
+
+	// updated it
+	c->data = roData;
```
If you make `->data` a `std::unique_ptr` and call `reset`, I think the cleanup logic is simpler.

davidchisnall commented on configuration_broker/config_broker/config_broker.h:
```diff
@@ -0,0 +1,145 @@
+{
+	uint16_t   id;             // id for the capability, assigned on first use
+	size_t     size;           // size of the item
+	uint32_t   updateInterval; // Min MS between updates
```
Is this ticks or mS?  The docs are a bit inconsistent.

PhilDay-CT replied:
Here, where the capability is being defined, its mS.     In the internal record of the broker its stored as ticks (in minTicks) to avoid converting it each time.   Can you point me to where the docs are inconsistent so I can make it clearer ?

davidchisnall commented on configuration_broker/config_broker/config_broker.h:
```diff
@@ -0,0 +1,145 @@
+	DECLARE_AND_DEFINE_STATIC_SEALED_VALUE(                                    \
+	  struct {                                                                 \
+		  uint16_t   id;                                                       \
+		  size_t     notUsed1;                                                 \
+		  uint32_t   notUsed2;                                                 \
```
Not used because they're always initialised to zero but set later?  Or do they not need to be there?

PhilDay-CT replied:
As per the preceding comment block its to keep the a common struct between all three capability types, so I can have a common unseal method in the broker.

davidchisnall commented on configuration_broker/config_broker/config_broker.h:
```diff
@@ -0,0 +1,145 @@
+	DECLARE_AND_DEFINE_STATIC_SEALED_VALUE(                                    \
+	  struct {                                                                 \
+		  uint16_t   id;                                                       \
+		  size_t     notUsed1;                                                 \
+		  uint32_t   notUsed2;                                                 \
```
Same question.

davidchisnall commented on configuration_broker/config_broker/config_broker.h:
```diff
@@ -0,0 +1,145 @@
+//
+struct ConfigItem
+{
+	const char            *name;         // name
+	uint32_t               version;      // version
+	void                  *data;         // value
+	std::atomic<uint32_t> *versionFutex; // Futex to wait for version change
```
Probably not important, but you could save a bit of memory if you moved `name`, `data`, and `versionFutex` into a separate structure and embedded this in `ConfigItemInternal` and just returned a single read-only capability to this.  This structure then becomes a pair of last-seen-version and the const pointer to the item information.

PhilDay-CT replied:
My thinking in separating out the version value from the futex was to avoid that the consumer ever has to read the version value from there outside of the context of the wait.   Even as an atomic it seems like there was potential for a race condition in there between threads in different consumers if I leave it up to them when to read the version value.  With the code as it is each consumer gets their own stack object with a consistent version and data pointer, where although *data and *versionFutex are shared they are readonly.  Even if there is an update between this struct being returned and the called reading it  data and version are always consistent.

davidchisnall commented on configuration_broker/doc/broker.puml:
```diff
@@ -0,0 +1,58 @@
+@startuml
```
What is this file?

rmn30 replied:
Looks like it is the source for the `svg` in Plant UML.

PhilDay-CT replied:
Its the source file for the diagram in the readMe - using https://plantuml.com/.

We use it a lot in our docs as in means you can easily keep the source of diagrams in the repo and the syntax is pretty easy to read and update.   We haven't found anything else that covers the same scope.

I can add a comment in there explaining ?

davidchisnall replied:
We use Mermaid for the same thing, GitHub renders it as inline diagram blocks in Markdown.

PhilDay-CT replied:
I can try converting it to Mermaid if you like, I just found it a bit limited compared to plantUML in the past,

PhilDay-CT replied:
I can try and covert to Mermaid if you want - I've always found it rather limited compared to plantuml

davidchisnall commented on configuration_broker/logger/logger.h:
```diff
@@ -0,0 +1,33 @@
+
+#pragma once
+
+#include <cstdlib>
```
Probably should be stdlib.h if this is expected to be included by C or C++.

davidchisnall commented on configuration_broker/parser/parser.cc:
```diff
@@ -0,0 +1,192 @@
+	parsed = parsed && get_enum<logLevel>(json, "level", &config->level);
+
+	// Free any heap the parser might have left allocated
+	auto heap_freed = heap_free_all(MALLOC_CAPABILITY);
```
I thought this compartment wasn't allowed to allocate or free memory?

PhilDay-CT replied:
Ideally it wouldn't - but as per the comment in parser.h using magic_enum to convert from string values in JSON requires new(), so the next best thing seems to be to force a tidy up if it left anything around (from what can see it never does).   I think it makes a good story about using compartments to isolate code that its not always possible to clean up first.  

In the next version I'm going to move each parser into its own compartment and give them each a very small heap quota so that even if one parser did get implemented without this it wouldn't affect any the parsing of any other items.

simonarnell commented on configuration_broker/README.md:
```diff
@@ -0,0 +1,209 @@
+Safe Configuration Management
+=============================
+
+Contributed by ConfiguredThings Ltd
```
Missing space between "Configured" and "Things"

simonarnell commented on configuration_broker/README.md:
```diff
@@ -0,0 +1,209 @@
+
+Contributed by ConfiguredThings Ltd
+
+All system rely on some form of configuration data, and the configuration interface is a significant part of the attack surface.
```
"system" or systems?

nwf replied:
```suggestion
All systems rely on some form of configuration data, and the configuration interface is a significant part of the attack surface.
```

simonarnell commented on configuration_broker/README.md:
```diff
@@ -0,0 +1,209 @@
+This avoids any timing issues during system startup.
+
+In the demo new values are provided as serialised JSON strings, with each item coming from a specific topic from a (not included) MQTT Broker.
+The Provider is in effect a simply an authorised mapping between the topics and configuration items.
```
> a simply an authorised 

should probably be "simply an authorised"

simonarnell commented on configuration_broker/README.md:
- Cheriot -> CHERIoT
- "runs as stateless method" -> "runs as a stateless method"
- "ensure" -> "ensures"

simonarnell commented on configuration_broker/README.md:
"stared" -> "started"

nwf commented on configuration_broker/README.md:
```suggestion
Using CHERIoT each parser runs as stateless method (using heap controls) in a sandbox compartment that ensure that any issues are contained to failing only the current parse operation.
```

nwf commented on configuration_broker/README.md:
```suggestion
Parsers are given a static sealed capability for each item type they are allowed to parse, which includes two properties of the item.
```

nwf commented on configuration_broker/README.md:
```suggestion
* The minimum interval in milliseconds between updates.
```

nwf commented on configuration_broker/config_broker/config_broker.h:
```diff
@@ -0,0 +1,145 @@
+//
+struct ConfigToken
+{
+	uint16_t   id;             // id for the capability, assigned on first use
```
AFAICT `id`s are unused other than to be assigned?

You could make this an `InternalConfigItem*`, if you wanted, to avoid the need to use `find_or_create_config` every time; the auditor can assert that it's `NULL` in all constructed instances.

If not, consider permuting `id` to be after `size` and `updateInterval` to avoid padding bytes.

PhilDay-CT replied:
yeah its not currently used - it was there for the same reason as the one in the malloc capabaility, to be able to track different instances of equivalent capabilities (e.g. the separate capabilites in each consumer to read the logger confi value).  Maybe I'll just remove it for now. 

The padding will still be there though as that's to give a common struct for all three capability types, and the name has to be the last item as its variable length.

nwf commented on configuration_broker/consumers/consumer1.cc:
```diff
@@ -0,0 +1,196 @@
+DEFINE_READ_CONFIG_CAPABILITY(LOGGER_CONFIG)
+
+// Expose debugging features unconditionally for this compartment.
+using Debug = ConditionalDebug<true, "Consumer #1">;
```
It's extremely nitpicky, but might whitespace between the `"` and the `>` un-confuse GitHub's approximate C++ parser?

nwf commented on configuration_broker/parser/parser.cc:
```diff
@@ -0,0 +1,192 @@
+// Parsing emums with magic_enum needs heap access, so
+// we can't use the following macros to configure the
+// compartment with no heap access. Instead we call
+// heap_free_all() at the end of each parser callback.
```
I think that's fairly expensive, as it's going to walk all objects in the heap to decide if they're allocated by this malloc capability.

PhilDay-CT replied:
I'll rework it tto check the quota remaining before and after and only call free is anything has been allocated.

nwf commented on configuration_broker/config_broker/config_broker.cc:
```diff
@@ -0,0 +1,343 @@
+		return -EPERM;
+	}
+
+	//
+	// We need to load all the parsers before we can
+	// accept any updates. Rather than have a thread in
+	// the parser compartment that runs to do that and
+	// then has to hang around (as threads can't exit) we
+	// use the first thread that tries to load a value to
+	// make a cross-comparement call to parserInit()
+	//
+	static bool ParsersLoaded = false;
+
+	if (!ParsersLoaded)
+	{
+		parserInit();
+		ParsersLoaded = true;
+	}
```
This is probably the best that's easily doable right now, but we should get an initialization framework in the base RTOS.  Something that runs a toposorted list of functions-in-compartments just after the loader has finished setting things up and before handing over the core to the switcher and scheduler.

PhilDay-CT replied:
Created this to track it: https://github.com/microsoft/cheriot-rtos/issues/275

nwf commented on configuration_broker/config_broker/config_broker.cc:
```diff
@@ -0,0 +1,343 @@
+	}
+
+	// Create a write only Capability to pass to the parser so
+	// that it can't capture or read from it
```
This, which clears `MC`, also prevents storing capabilities, which might be worth mentioning.

nwf commented on configuration_broker/config_broker/config_broker.cc:
```diff
@@ -0,0 +1,343 @@
+	woNewData.permissions() &= {CHERI::Permission::Store};
+
+	// Create a read only Capability of the json string to pass
+	// to the parser so that it can't capture or change it
```
Similarly clears `MC`

PhilDay-CT replied:
If its read only isn't it implicit that you can;t sore anything in it ?

nwf-msr replied:
Yes, but clearing MC also means you can't read capabilities through it.

nwf reviewed (commented):
This remains a delight to see and thank you so much for taking the time.  Still making my way through, but some initial suggestions (do feel free to ignore any/all).

- `configuration_broker/parser` is a bit of a mashup between reusable components and things specific to this example deployment.  Could it be worth splitting that?

rmn30 commented on configuration_broker/README.md:
```diff
@@ -0,0 +1,209 @@
+## Overview
+
+Each item of configuration data has a name, a value, and a version.
+New values are supplied as serialised JSON and parsed into a corresponding data strut. 
```
```suggestion
New values are supplied as serialised JSON and parsed into a corresponding data structure. 
```
or perhaps `C struct`?

simonarnell commented on configuration_broker/xmake.lua:
```diff
@@ -0,0 +1,105 @@
+-- Copyright Configred Things and CHERIoT Contributors.
```
```suggestion
-- Copyright Configred Things and CHERIoT Contributors.
++ Copyright Configured Things and CHERIoT Contributors.
```

PhilDay-CT committed:
Responding to Review Comments
 - Split parsers into seperate compartments
 - JSON parser moved to a library
 - Make use of namespaces
 - doc style comments where relevant
 - diagrams in readMe switched to mermaid
 - Make consumer threads exit
 - Various typos fixed

simonarnell commented on configuration_broker/README.md:
```suggestion
Consumer #1 is authorised to receive the RGB LED configuration.
```

PhilDay-CT committed:
remove LoadStoreCap from validated values

simonarnell commented:
@PhilDay-CT thanks for considering my comments, looks good to me.
